### ISSUE TITLE
IPv6 support

### ISSUE THREAD

**Issue by fengli79** [lyft/envoy#351]
Does envoy support IPv6? If not, is there a plan to support it?

**Comment by mattklein123** [comment]
It took a lot to have this issue opened than I thought it would. 😉 

It's not currently supported. It wouldn't be difficult to support, but someone needs to do the work.

**Comment by jamessynge** [comment]
Can you provide any pointers to where support is missing (e.g. int fields for addresses, or parsing of only IPv4 addresses)?

**Comment by fengli79** [comment]
Here's an example:
https://github.com/lyft/envoy/blob/c36bb76e304ab4c5584a07303e2953a170888319/source/common/network/listener_impl.cc#L104

And if you search AF_INET6 in the repo, there's no reference.

**Comment by mattklein123** [comment]
This is somewhat related to https://github.com/lyft/envoy/issues/390 which I am going to fix soon (next week most likely).

In general, it depends on where we need/want IPv6 support. There will be a small number of changes sprinkled through the code base, and not all of them are necessarily needed at once:
1) Add IPv6 listener support
2) Add IPv6 proxy proto support
3) Add IPv6 address parsing support for upstream connections, logging, etc.
4) Add IPv6 IP list support (for white lists, TCP proxy routing, etc.)

I would like to wait until https://github.com/lyft/envoy/issues/390 is done before taking this on. Also if we could get crisp about which scenarios people want explicitly, that would be great, and I can point you in the right direction if you want to work on this.

**Comment by mattklein123** [comment]
https://github.com/lyft/envoy/issues/390 is done. This should be pretty straightforward at this point if someone wants to take this on.

**Comment by jamessynge** [comment]
I'll take a look on Monday.

**Comment by jamessynge** [comment]
FYI, I'm going to start with adding some tests, e.g. for common/network/listener, that use IPv6 loopback rather than IPv4 loopback.

**Comment by jamessynge** [comment]
For "class Ipv6" (analagous to Network::Address::Ipv4), is it OK to use a struct such as in6_addr, as the return type for the address, or is there a preference to avoid using platform specific headers in this code?

**Comment by mattklein123** [comment]
IIRC we don't even need that right now. The only place that would be required is if/when we implement IPv6 address/mask matching in IpList, which isn't necessarily required for v1 of IPv6 support. Optimally, it would be nice to not expose system types (just thinking about eventual NT port), but given how infrequently that data will be used, it's fine if you want to do that for now.

**Comment by jamessynge** [comment]
ProxyProtocol::ActiveConnection::onReadWorker reads an IPv4 range from somewhere (apparently not the json config file), and then passes it to Network::IpList, which only handles IPv4. Having reviewed the protocol, it looks like it should also be extended to support IPv6. Agreed?

More generally, I suspect that I should convert IpList to handle IPv4 and IPv6 addresses, though perhaps with some option to limit which is allowed on any one call.

**Comment by mattklein123** [comment]
> ProxyProtocol::ActiveConnection::onReadWorker reads an IPv4 range from somewhere (apparently not the json config file), and then passes it to Network::IpList, which only handles IPv4. Having reviewed the protocol, it looks like it should also be extended to support IPv6. Agreed?

Sorry not following. Did you link the wrong function? The proxy proto code just implements http://www.haproxy.org/download/1.5/doc/proxy-protocol.txt. It would be nice to make this work for IPv6 at some point.

> More generally, I suspect that I should convert IpList to handle IPv4 and IPv6 addresses, though perhaps with some option to limit which is allowed on any one call.

Yes, IpList is currently used in both tcp_proxy routing as well as client_ssl auth white listing. It would be great to make this work for IPv6 in both cases.

In general though, I don't think you need to do a single giant PR to add IPV6 support everywhere. (And in fact I would recommend you don't). There are a bunch of separate places where we can add IPv6 support and test:

- Downstream listener IPv6
- Upstream IPv6
- IpList IPv6
- ProxyProto IPv6

**Comment by jamessynge** [comment]
Agreed, I was planning that the current PR would be separate from adding support elsewhere.

Ignore what I said about ProxyProtocol, I was a bit confused when I started writing that.

**Comment by mattklein123** [comment]
@jamessynge sounds great. FWIW, I think IpList to IPv6 is probably pretty low priority, but up to you on order you want to tackle everything.

**Comment by jamessynge** [comment]
Priorities are helpful. I've obviously started on IpRange, but where do you suggest I focus? Downstream listener and DNS AAAA record handling?

**Comment by mattklein123** [comment]
I think the biggest use case of IPv6 is probably downstream listener. Many/most people are going to end up doing IPv6 -> IPv4. So if we start there and make sure all access logging, etc. is working I think that's a great start. (We also need to think through config in terms of whether we want to allow dual bind or IPv4 and IPv6 only listeners, etc.).

Once downstream listener is done, then I would probably do all the other downstream associated stuff like ProxyProto and IPList. 

I would probably do upstream/DNS last.

(I don't really have a strong preference on any of this though if you need get specific flows working).

**Comment by jamessynge** [comment]
I'll be at Next 2017 for all of next week, but if I have some free time I'll start on the downstream listener.
Meanwhile, I've been completing IpRange, which I'll send for review RSN.

**Comment by jamessynge** [comment]
Matt, do you want to assign this to me?

**Comment by mattklein123** [comment]
I will add you are read contributor then I can assign to you. Look out for the accept request.

**Comment by jamessynge** [comment]
@mattklein123 - Regarding this code:

```
TcpListenSocket::TcpListenSocket(uint32_t port, bool bind_to_port) {
  // TODO(mattklein123): IPv6 support.
  local_address_.reset(new Address::Ipv4Instance(port));
  fd_ = local_address_->socket(Address::SocketType::Stream);
```

What would be your thought w.r.t. adding IPv6 support? Add a new parameter to specify Address::IpVersion? Or just remove this method, and require user to use the previous method which takes an address, and hence already supports IPv6.

**Comment by mattklein123** [comment]
@jamessynge I commented on it in the previous @wattli  review, but I would really prefer that we get rid of the port based constructors here entirely, and just keep the ones that take AddressPtr. Then we can deal with how things are configured independently.

I still have some general questions on what types of configurations we want to allow (i.e., dual bind, single bind, etc.).

**Comment by jamessynge** [comment]
According to [this answer on serverfault](http://serverfault.com/a/39561), dual bind is a bit complicated (I was ignorant about its behavior until I read up on it this morning). The part that was particularly interesting was that on a Linux host configured for dual bind (the default), if a server program binds to "::80", and then an IPv4 connection from 10.1.1.3 is established to this host's port 80, the client's address will be mapped to IPv6 as ::ffff:10.1.1.3, which can confuse matters. It turns out that when I look at our internal proxy config, we depend on the dual stack mode (explicitly enabled on each socket with sockopt IPV6_V6ONLY), and map addresses (client and server) back to IPv4 when appropriate (i.e. if the client initiated an IPv4 connection, we treat the server and client addresses as IPv4, even though they are apparently reported as IPv6 addresses by accept and getpeername).

I'm assuming there is not already an option for dual bind because Envoy has not previously had IPv6 support. Going forward, we need to consider how we treat "0.0.0.0" and "::" when binding.

- As independent (i.e. we force IPV6_V6ONLY to 1 when binding "::").
- As equivalent (i.e. we collapse such configs into one if possible, if we see either one, we bind to "::" with IPV6_V6ONLY set to 0)
- As equivalent IFF the user has indicated they should be equivalent, either by specifying it with a new listener config option, or implicitly with a new addressing scheme; e.g. "tcp://:80" (no address at all) or "tcp://any-address:80".

I suspect the most straightforward choice would be to introduce a new config option on listeners, ipv6_v6only, default true. Then a user wishing to upgrade from IPv4 only with address "tcp://0.0.0.0:80" to dual bind would change the address to "tcp://[::]:80" and set ipv6_v6only=false in the config.

Thoughts?

**Comment by mattklein123** [comment]
> I suspect the most straightforward choice would be to introduce a new config option on listeners, ipv6_v6only, default true. Then a user wishing to upgrade from IPv4 only with address "tcp://0.0.0.0:80" to dual bind would change the address to "tcp://[::]:80" and set ipv6_v6only=false in the config.

I agree. I'm in favor or having dual bind not be the default. I wonder if we even need to support dual bind initially or can just make people setup multiple listeners.

**Comment by jamessynge** [comment]
I can't say that I'm inclined towards multiple listeners as the filters list needs to be repeated, and IIUC the configs can't readily be shared. Is that so?

**Comment by mattklein123** [comment]
Currently, no, they can't be shared easily at the JSON level. With that said, I think most people with any substantial deployment will be machine generating the configs so not sure it matters. In any case, I'm not opposed to having dual bind be an option, was just trying to make it easier for you if you want to do that as a follow up.

**Comment by hennna** [comment]
@mattklein123 I'll be continuing the Ipv6 work from @jamessynge.

I think a good place to start would be to add ipv6 address parsing to Utility::resolveUrl: https://github.com/lyft/envoy/blob/master/source/common/network/utility.cc#L89. For pipes, just s/unix/pipe. 

The next step would be to incrementally updates tests to include ipv6 addresses. Thoughts?

**Comment by mattklein123** [comment]
@hennna if you want to start from a test first perspective, I think that's fine, but I would decide whether you want to start from a unit test first perspective or an integration test first perspective. @jamessynge has been making good progress basically tackling various pieces and making sure there are unit tests for each piece. 

If you want to start from an integration test perspective, the simplest end to end test is probablyhttps://github.com/lyft/envoy/blob/master/test/integration/integration_test.cc#L13

This test requires no upstream support in Envoy. I would recommend that you work backward and get that test working in IPv6, adding unit tests along the way as you alter individual components. For that test to work with IPv6 only you will need IPv6 listener support as well as IPv6 fake client support. It will also allow us to start figuring out how to parameterize the integration tests for IPv6.

re: Utility::resolveUrl, I think the comment in there is out of data a bit re: ip:// and pipe://. I would leave it as tcp:// for now since eventually we will want to also support udp:// but we can deal with that later. But yes, you will need to modify that function at some point to support IPv6. 

However I think I would start with the echo test.

**Comment by hennna** [comment]
Yes that sounds good. I can start looking at the echo test.

I'm still looking into bazel and environment variables to see how users can have the option to run tests for ipv4 only, ipv6 only or both.

**Comment by hennna** [comment]
The following tests currently fail in an Ipv6 only environment:
test/common/network:address_impl_test
	Ipv4InstanceTest.SocketBindAndConnect
test/common/network:connection_impl_test
test/common/network:dns_impl_test
test/common/network:listener_impl_test
test/common/network:proxy_protocol_test
test/common/network:utility_test 
	NetworkUtility.getLocalAddress
test/common/event:file_event_impl_test
test/common/ssl:connection_impl_test
test/integration/… (all tests)
test/server/http:admin_test

**Comment by mattklein123** [comment]
Closing this out as I believe are done with basic IPv6 support. Awesome work @hennna !
